### PR TITLE
shimv2: improve error reporting when the monitor thread fails

### PR DESCRIPTION
**Pull request by fidencio**
## Problem
When the monitor thread fails (Vmm or agent  failed to respond). The monitor notifies that something that an error happen and mark the agent as dead. After that, depending on when the notification happen this can result on not well propagated errors (or not event report errors). The main reason is a concurrency problem. 

1. Multiple shim api calls can happen at the same time.
2. The container state is not thread safe, if an error is propagated  and also the container was marked as stopped (after an unexpected error), it can hapend that some api calls like Wait or State just return that the status of the container changed, but the real reason was en error that was not seen because.
3. If an error is well propagated only the most inner error is shown that does not include a lot of information.

## Solution

Improve error reporting:
- virtcontainers: Add errors package
- errors: not modify error if is not grpc type
- errors: use virtcontainers.errors for monitor failing
- monitor: Stop watcher on error

Improve state update of shim container struct
- shim: improve concurrency for container state
- monitor: Stop watcher on error

Example of error trace from kubeadm
```
Normal Scheduled 30s default-scheduler Successfully assigned default/config-env-test-pod to jcvenega-kata-clh-debug Normal Pulling 26s kubelet Pulling image "quay.io/prometheus/busybox:latest" Normal Pulled 26s kubelet Successfully pulled image "quay.io/prometheus/busybox:latest" in 431.185669ms Normal Created 24s kubelet Created container test-container Warning Failed 21s kubelet Error: failed to create containerd task: failed to create shim: Cause: MYERROR

Error trace: 0. MYERROR

Failed to check if Cloud Hypervisor is working
Monitor has failed to watch hypervisor
Monitor marked the agent as dead
Agent was marked dead
ttrpc: closed
Failed to send request grpc.CreateContainerRequest
Agent failed to create container
Failed to create container
Failed to create container in sandbox
shimv2 create failed
The Create() endpoint from kata shim failed
Stack: 0. github.com/kata-containers/kata-containers/src/runtime/virtcontainers/errors.ErrorContext /mnt/go/src/github.com/kata-containers/kata-containers/src/runtime/virtcontainers/errors/errors.go +33

github.com/kata-containers/kata-containers/src/runtime/virtcontainers.(*cloudHypervisor).Check /mnt/go/src/github.com/kata-containers/kata-containers/src/runtime/virtcontainers/clh.go +791
github.com/kata-containers/kata-containers/src/runtime/virtcontainers.(*monitor).watchHypervisor /mnt/go/src/github.com/kata-containers/kata-containers/src/runtime/virtcontainers/monitor.go +153
github.com/kata-containers/kata-containers/src/runtime/virtcontainers.(*monitor).newWatcher.func1 /mnt/go/src/github.com/kata-containers/kata-containers/src/runtime/virtcontainers/monitor.go +64
runtime.goexit /usr/local/go/src/runtime/asm_amd64.s +1581: unknown Warning FailedKillPod 19s (x2 over 20s) kubelet error killing pod: failed to "KillPodSandbox" for "892b6bf8-04e4-477b-92a9-86154e24bcdc" with KillPodSandboxError: "rpc error: code = Unknown desc = failed to stop sandbox container "f0d57bfe72f4d69a5f58f09d4beac291c560a7cbf2467feb7cb199f4367a2df3" in "SANDBOX_READY" state: failed to kill sandbox container: Container.wait\n\tCause: Failed to wait for process from agent\n\tCause: Failed to send request grpc.WaitProcessRequest\n\tCause: ttrpc: closed\n\tCause: Agent was marked dead\n\tCause: Monitor marked the agent as dead\n\tCause: Monitor has failed to watch hypervisor\n\tCause: Failed to check if Cloud Hypervisor is working\n\tCause: MYERROR\n\tCause: Agent was marked dead\n\tCause: Monitor marked the agent as dead\n\tCause: Monitor has failed to watch hypervisor\n\tCause: Failed to check if Cloud Hypervisor is working\n\tCause: MYERROR: unknown"
```

### CODE REVIEW

**Comment by fidencio** [comment]
This is **not** my work, this is a rebase of the amazing work @jcvenegas has done on #3275.
@liubin @jodh-intel, if I could get your attention here, that would be amazing.

**Comment by c3d** [comment]
Thanks @fidencio for reviving this. This looks really useful.

**Review by jodh-intel — DISMISSED** [review]
Thanks @fidencio and @jcvenegas !

lgtm

File: src/runtime/pkg/containerd-shim-v2/service_test.go [hunk 1]
```diff
@@ -36,8 +36,9 @@ func newService(id string) (*service, error) {
 }
 
 func TestServiceCreate(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (fidencio) [reply src/runtime/pkg/containerd-shim-v2/service_test.go:38]: I tried to address this by going through the majority of the commits and re-writing their commit messages in a way that could be cleaner (or not, let's see).

**Comment by fidencio** [comment]
So, my 2 cents about this series.  IMHO, this is a huge improvement but there are patches here and there that could've been squashed together ... and there are some parts that are not that much clear to me.

I'm going to propose to **hold** this one till 2.4.0-rc0 is cut, and we revisit this at the beginning of 2.5 release cycles.  This gives us more time for testing the content that will be in the release.

For now I'm adding a do-not-merge label, but I'd be happy to remove it in case someone else has a different opinion.

**Comment by c3d** [comment]
> So, my 2 cents about this series. IMHO, this is a huge improvement but there are patches here and there that could've been squashed together ... and there are some parts that are not that much clear to me.
> 
> I'm going to propose to **hold** this one till 2.4.0-rc0 is cut, and we revisit this at the beginning of 2.5 release cycles. This gives us more time for testing the content that will be in the release.
> 
> For now I'm adding a do-not-merge label, but I'd be happy to remove it in case someone else has a different opinion.

We need a "do-not-merge-until-nexrt-release-cycle" label ;-) 

Just curious, why did you open a separate PR from #3275? Do we really need the two PRs?

**Comment by fidencio** [comment]
> We need a "do-not-merge-until-nexrt-release-cycle" label ;-)
> 
> Just curious, why did you open a separate PR from #3275? Do we really need the two PRs?

Carlos won't be working on this one, and for some reason I never managed to be able to force-push into someone's else PR.

**Review by jodh-intel — APPROVED** [review]
Thanks @fidencio.

lgtm - could you rebase to resolve the conflict though? :)

**Comment by jodh-intel** [comment]
@fidencio - could you rebase to resolve the conflict please.

**Comment by jodh-intel** [comment]
Re-ping @fidencio - once this is rebased we can land it! ;)

**Comment by arronwy** [comment]
Can one of the admins verify this patch?

**Comment by jodh-intel** [comment]
This one still needs a rebase.

**Pull request closed without merging**